units/event_log_tests.cc

```diff
@@ -0,0 +1,213 @@
+        timestamp_{"timestamp", timestamp},
+        hash_value_{"hash", hash_value},
+        duration_{"duration", duration},
+        attributes_{&timestamp_, &hash_value_, &duration_} {}
+
+  const std::vector<Attribute *> &GetAttributes() override {
+    return attributes_;
+  }
+  size_t GetNumAttributes() const override { return attributes_.size(); }
+
+ private:
+  Int64Attr timestamp_;
+  Int64Attr hash_value_;
+  Int64Attr duration_;
+  std::vector<Attribute *> attributes_;
+};
+
+std::string ValueToCSVString(const std::string &value) { return value; }
+
+std::string ValueToCSVString(const int64_t value) {
+  return std::to_string(value);
+}
+
+std::string ValueToCSVString(const std::vector<int64_t> &values) {
+  std::stringstream csv_string;
+  csv_string << "\"[";
+  size_t e = values.size();
+  for (size_t i = 0; i != e; ++i) {
+    const char *delimiter = i < e - 1 ? "," : "";
+    csv_string << std::to_string(values[i]) + delimiter;
+  }
+  csv_string << "]\"";
+  return csv_string.str();
+}
+
+// Takes an `Event` instance as an input and generates a csv string containing
+// `event`'s name and attribute values.
+std::string EventToCSVString(Event &event) {
+  const std::vector<Attribute *> &attributes = event.GetAttributes();
+
+  std::stringstream csv_str;
+  csv_str << event.GetEventName();
+  csv_str << ",";
+  size_t e;
+  for (size_t i = 0, e = attributes.size(); i != e; ++i) {
+    switch (attributes[i]->GetValueType()) {
+      case ValueType::Int64: {
+        csv_str << ValueToCSVString(
+            attributes[i]->cast<Int64Attr>()->GetValue());
+        break;
+      }
+      case ValueType::String: {
+        csv_str << ValueToCSVString(
+            attributes[i]->cast<StringAttr>()->GetValue());
+        break;
+      }
+      case ValueType::VectorInt64: {
+        csv_str << ValueToCSVString(
+            attributes[i]->cast<VectorInt64Attr>()->GetValue());
+        break;
+      }
+    }
+    if (i + 1 != e) csv_str << ",";
+  }
+  return csv_str.str();
+}
+
+TEST(Event, AttributeCreation) {
+  const Int64Attr timestamp("timestamp", 1601314733266018318);
+  const StringAttr state("state", "1");
+  const VectorInt64Attr pipeline("pipeline",
+                                 {0x67d6fd0aaa78a6d8, 0x67d390249c2f20ce});
+  EXPECT_EQ(timestamp.GetName(), "timestamp");
+  EXPECT_EQ(timestamp.GetValue(), 1601314733266018318);
+  EXPECT_EQ(state.GetName(), "state");
+  EXPECT_EQ(state.GetValue(), "1");
+  EXPECT_EQ(pipeline.GetName(), "pipeline");
+  ASSERT_EQ(pipeline.GetValue().size(), 2);
+  EXPECT_EQ(pipeline.GetValue()[0], 0x67d6fd0aaa78a6d8);
+  EXPECT_EQ(pipeline.GetValue()[1], 0x67d390249c2f20ce);
+}
+
+TEST(Event, CreateShaderModuleEventCreation) {
+  CreateShaderModuleEvent compile_event("compile_time", 1601314733266018318,
+                                        0x67d6fd0aaa78a6d8, 926318,
+                                        LogLevel::Low);
+  ASSERT_EQ(compile_event.GetNumAttributes(), 3);
```

Comment:
Here `EXPECT_EQ` is fine because there is no other check that would otherwise fail.